EIPS/eip-5450.md

```diff
@@ -13,51 +13,103 @@ requires: 3540, 3670, 4200, 4750
 
-*Data stack height* here refers to the number of stack values accessible by this function, i.e. it does not take into account values of caller functions' frames (but does include this function's inputs). Note that validation procedure does not require actual data stack implementation, but only to keep track of its height. Current height value starts at `type[code_section_index].inputs` (number of inputs of this function) at function entry and is updated at each instruction.
+*Operand stack height* here refers to the number of stack values accessible by this function, i.e. it does not take into account values of caller functions' frames (but does include this function's inputs). Note that validation procedure does not require actual data stack implementation, but only to keep track of its height.
 
-At the same time the following properties are being verified:
+For each reachable instruction in the code the operand stack height is recorded.
+The first instruction has the recorded stack height equal to the number of inputs to the function type matching the code (`type[code_section_index].inputs`).
 
-1. For each reachable instruction in the section, data stack height is the same for all possible code paths going through this instruction.
-2. For each instruction, data stack always has enough items, i.e. stack underflow is invalid.
-3. For `CALLF` instruction, data stack has enough items to use as input arguments to a called function according to its type defined in the type section.
-4. For `RETF` instruction, data stack before executing it has exactly `n` items to use as output values returned from a function, where `n` is function's number of outputs according to its type defined in the type section.
-5. `max_stack_height` does not exceed 1023.
-6. The maximum data stack height matches the corresponding code section's `max_stack_height` within the type section body.
-7. No unreachable instructions exist in the code section.
+The FIFO queue *worklist* is used for tracking "to be inspected" instructions. Initially, it contains the first instruction.
 
-To examine every reachable code path, validation needs to traverse every instruction in order, while also following each non-conditional jump, and following both possible branches for each conditional jump. After completing this, verify each instruction was visited at least once. Fail if any instructions were not visited as this invalidates 7).
+While *worklist* is not empty, dequeue an instruction and:
 
-The complexity of this traversal is linear in the number of instructions, because each code path is examined only once, and property 1 guarantees no loops in the validation.
+1. Determine the effect the instruction has on the operand stack:
+   1. **Check** if the recorded stack height satisfies the instruction requirements. Specifically:
+      - for `CALLF` instruction the required stack height is the number of inputs of the called function according to its type defined in the type section.
```

Comment:
I guess the wording of the rule ``1.1.CALLF`` should reflect that, instead of saying at ``CALLF`` sight the  ``..required stack height is the number of the inputs of the called function..``, I think it should say ``..the required stack height is at least the number of inputs of the called function..``

Reply:
```suggestion
      - for `CALLF` instruction the required stack height is at least the number of inputs of the called function according to its type defined in the type section.
```

Reply:
Changed by avoiding "required stack height" term (I use it internally and it slipped into the spec).